src/dotnet/commands/dotnet-add/dotnet-add-package/Program.cs

```diff
@@ -0,0 +1,174 @@
+// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Microsoft.Build.Evaluation;
+using Microsoft.DotNet.Cli;
+using Microsoft.DotNet.Cli.CommandLine;
+using Microsoft.DotNet.Cli.Utils;
+using Microsoft.DotNet.Tools.Common;
+using Microsoft.DotNet.Tools.MSBuild;
+using Microsoft.DotNet.Tools.NuGet;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.Tools.Add.PackageReference
+{
+    internal class AddPackageReferenceCommand : DotNetSubCommandBase
+    {
+        private CommandOption _versionOption;
+        private CommandOption _frameworkOption;
+        private CommandOption _noRestoreOption;
+        private CommandOption _sourceOption;
+        private CommandOption _packageDirectoryOption;
+
+        public static DotNetSubCommandBase Create()
+        {
+            var command = new AddPackageReferenceCommand()
+            {
+                Name = "package",
+                FullName = LocalizableStrings.AppFullName,
+                Description = LocalizableStrings.AppDescription,
+                HandleRemainingArguments = true,
+                ArgumentSeparatorHelpText = LocalizableStrings.AppHelpText,
+            };
+
+            command.HelpOption("-h|--help");
+
+            command._versionOption = command.Option(
+                $"-v|--version <{LocalizableStrings.CmdVersion}>",
+                LocalizableStrings.CmdVersionDescription,
+                CommandOptionType.SingleValue);
+
+            command._frameworkOption = command.Option(
+               $"-f|--framework <{LocalizableStrings.CmdFramework}>",
+               LocalizableStrings.CmdFrameworkDescription,
+               CommandOptionType.SingleValue);
+
+            command._noRestoreOption = command.Option(
+                "-n|--no-restore ",
+               LocalizableStrings.CmdNoRestoreDescription,
+               CommandOptionType.NoValue);
+
+            command._sourceOption = command.Option(
+                $"-s|--source <{LocalizableStrings.CmdSource}>",
+                LocalizableStrings.CmdSourceDescription,
+                CommandOptionType.SingleValue);
+
+            command._packageDirectoryOption = command.Option(
+                $"--package-directory <{LocalizableStrings.CmdPackageDirectory}>",
+                LocalizableStrings.CmdPackageDirectoryDescription,
+                CommandOptionType.SingleValue);
+
+            return command;
+        }
+
+        public override int Run(string fileOrDirectory)
+        {
+            var projects = new ProjectCollection();
+            var msbuildProj = MsbuildProject.FromFileOrDirectory(projects, fileOrDirectory);
```

Comment:
Should you move this to below te RemainingArguments check? This is potentially costly for the user to wait for and then fail.